File: src/test/java/br/com/blz/testjava/application/product/delete/DeleteProductUseCaseTest.java
package br.com.blz.testjava.application.product.delete;

import br.com.blz.testjava.application.UseCaseTest;
import br.com.blz.testjava.application.product.InventoryInput;
import br.com.blz.testjava.application.product.create.CreateProductCommand;
import br.com.blz.testjava.domain.product.ProductGateway;
import br.com.blz.testjava.domain.product.Sku;
import br.com.blz.testjava.domain.product.Warehouse;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.Mockito;

import java.util.List;

import static java.util.Arrays.asList;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.doNothing;
import static org.mockito.Mockito.times;

public class DeleteProductUseCaseTest extends UseCaseTest {

    @InjectMocks
    private DefaultDeleteProductUseCase useCase;

    @Mock
    private ProductGateway productGateway;

    @Override
    protected List<Object> getMocks() {
        return List.of(productGateway);
    }

    @Test
    public void givenAValidId_whenCallsDeleteProduct_shouldBeOK() {
        final var newProduct = CreateProductCommand.with(43264L,
            "L'Oréal Professionnel Expert Absolut Repair Cortex Lipidium - Máscara de Reconstrução 500g",
            InventoryInput.with(asList(Warehouse.newWarehouse("SP", 12, "ECOMMERCE"),
                Warehouse.newWarehouse("MOEMA", 3, "PHYSICAL_STORE"))));
        final var expectedSku = Sku.from(newProduct.sku());

        doNothing().when(productGateway).deleteBySku(eq(expectedSku));

        Assertions.assertDoesNotThrow(() -> useCase.execute(expectedSku.getValue()));

        Mockito.verify(productGateway, times(1)).deleteBySku(eq(expectedSku));
    }
}

File: src/test/java/br/com/blz/testjava/domain/usecase/DeleteProductUseCaseTest.kt
package br.com.blz.testjava.domain.usecase

import br.com.blz.testjava.domain.exception.NotFoundException
import br.com.blz.testjava.domain.repository.ProductRepository
import br.com.blz.testjava.domain.repository.WarehouseRepository
import br.com.blz.testjava.factory.MockFactory
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertDoesNotThrow
import org.junit.jupiter.api.assertThrows
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.Mockito.doReturn
import org.mockito.junit.jupiter.MockitoExtension

@ExtendWith(MockitoExtension::class)
class DeleteProductUseCaseTest {
  @Mock private lateinit var productRepository: ProductRepository
  @Mock private lateinit var warehouseRepository: WarehouseRepository

  @InjectMocks private lateinit var deleteProductUseCase: DeleteProductUseCase

  @Test
  fun `Should delete product`() {
    val product = MockFactory.createProduct()
    doReturn(product).`when`(productRepository).findById(product.sku.toString())

    assertDoesNotThrow {
      deleteProductUseCase.execute(product.sku.toString())
    }
  }

  @Test
  fun `Should throw not found exception when sku is not found`() {
    val product = MockFactory.createProduct()
    assertThrows<NotFoundException> {
      deleteProductUseCase.execute(product.sku.toString())
    }
  }

}
